resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java

```diff
@@ -65,6 +71,22 @@ public void operationComplete(Future<AddressedEnvelope<DnsResponse, InetSocketAd
                     }
                 }
             };
+    private static final RuntimeException NXDOMAIN_QUERY_FAILED_EXCEPTION = ThrowableUtil.unknownStackTrace(
+            new RuntimeException("No answer found and NXDOMAIN response code returned"),
+            DnsNameResolverContext.class,
+            "onResponse(..)");
+    private static final RuntimeException CNAME_NOT_FOUND_QUERY_FAILED_EXCEPTION = ThrowableUtil.unknownStackTrace(
+            new RuntimeException("No matching CNAME record found"),
+            DnsNameResolverContext.class,
+            "onResponseCNAME(..)");
+    private static final RuntimeException NO_MATCHING_RECORD_QUERY_FAILED_EXCEPTION = ThrowableUtil.unknownStackTrace(
+            new RuntimeException("No matching record type record found"),
+            DnsNameResolverContext.class,
+            "onResponseAorAAAA(..)");
+    private static final RuntimeException UNRECOGNIZED_TYPE_QUERY_FAILED_EXCEPTION = ThrowableUtil.unknownStackTrace(
+            new RuntimeException("Response type was unrecognized"),
+            DnsNameResolverContext.class,
```

Comment:
I modified my original comment above ... I don't want to use `DnsNameResolverException` because it requires state and cannot be generated statically. I want to avoid creating exceptions dynamically because this will generate overhead even when `NoopDnsQueryLifecycleObserver` is in use. A "real" implementation of `DnsQueryLifecycleObserver` has enough context through the API to reconstruct what happened and the `Throwable` is more meant to provide some additional context ... it could also just be a `String`, or some enum.